protocol/contracts/beanstalk/silo/ConvertFacet.sol

```diff
@@ -81,29 +117,331 @@ contract ConvertFacet is Invariable, ReentrancyGuard {
+        returns (int96 toStem, uint256 fromAmount, uint256 toAmount, uint256 fromBdv, uint256 toBdv)
+    {
+        // require that input and output tokens be wells (Unripe not supported)
+        require(
+            LibWell.isWell(inputToken) || inputToken == C.BEAN,
+            "Convert: Input token must be Bean or a well"
+        );
+        require(
+            LibWell.isWell(outputToken) || outputToken == C.BEAN,
+            "Convert: Output token must be Bean or a well"
+        );
+
+        PipelineConvertData memory pipeData;
+        pipeData.user = LibTractor._user();
+
+        // mow input and output tokens:
+        LibSilo._mow(pipeData.user, inputToken);
+        LibSilo._mow(pipeData.user, outputToken);
+
+        // Calculate the maximum amount of tokens to withdraw
+        for (uint256 i = 0; i < stems.length; i++) {
+            fromAmount = fromAmount.add(amounts[i]);
+        }
+
+        // withdraw tokens from deposits and calculate the total grown stalk and bdv.
+        (pipeData.grownStalk, fromBdv) = _withdrawTokens(inputToken, stems, amounts, fromAmount);
+
+        // Store the capped overall deltaB, this limits the overall convert power for the block
+        pipeData.overallConvertCapacity = LibConvert.abs(LibWellMinting.overallCappedDeltaB());
+
+        // Store the pre-convert insta deltaB's both overall and for each well
+        pipeData.beforeOverallDeltaB = LibWellMinting.overallCurrentDeltaB();
+        pipeData.beforeInputTokenDeltaB = getCurrentDeltaB(inputToken);
+        pipeData.beforeOutputTokenDeltaB = getCurrentDeltaB(outputToken);
+
+        pipeData.beforeInputLpTokenSupply = IERC20(inputToken).totalSupply();
+        pipeData.beforeOutputLpTokenSupply = IERC20(outputToken).totalSupply();
+        pipeData.initialLpSupply = LibWellMinting.getLpSupply();
+
+        IERC20(inputToken).transfer(C.PIPELINE, fromAmount);
+        executeAdvancedFarmCalls(advancedFarmCalls);
+
+        // user MUST leave final assets in pipeline, allowing us to verify that the farm has been called successfully.
+        // this also let's us know how many assets to attempt to pull out of the final type
+        toAmount = transferTokensFromPipeline(outputToken);
+
+        // Calculate stalk penalty using start/finish deltaB of pools, and the capped deltaB is
+        // passed in to setup max convert power.
+        pipeData.stalkPenaltyBdv = prepareStalkPenaltyCalculation(
+            inputToken,
+            outputToken,
+            pipeData.beforeInputTokenDeltaB,
+            pipeData.beforeInputLpTokenSupply,
+            pipeData.beforeOutputTokenDeltaB,
+            pipeData.beforeOutputLpTokenSupply,
+            pipeData.beforeOverallDeltaB,
+            pipeData.overallConvertCapacity,
+            fromBdv,
+            pipeData.initialLpSupply
+        );
+
+        // Update grownStalk amount with penalty applied
+        pipeData.grownStalk = pipeData.grownStalk.sub(pipeData.stalkPenaltyBdv);
```

Comment:
shouldn't `stalkPenaltyBdv` be scaled by the grown stalk? i.e `stalkPenaltyBdv` * grownStalkPerBdv of deposit?

Reply:
Fixed here: https://github.com/BeanstalkFarms/Beanstalk/pull/835/commits/a3f3c5144b29523af060c693552dcee3b81bd671

Reply:
lgtm!